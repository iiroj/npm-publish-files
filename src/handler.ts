import fs from "fs";
import cbGlob, { IOptions as GlobOptions } from "glob";
import minimatch from "minimatch";
import packlist from "npm-packlist";
import path from "path";

const ensureDir = async (dir: string) => {
  try {
    const stats = await fs.promises.stat(dir);
    if (!stats.isDirectory()) {
      throw new Error(`${dir} exists but is not a directory!`);
    }
  } catch (err) {
    if (err.code === "ENOENT") {
      await fs.promises.mkdir(dir);
    } else {
      throw err;
    }
  }
};

const glob = (pattern: string, options: GlobOptions) =>
  new Promise<string[]>((resolve, reject) =>
    cbGlob(pattern, options, (err, matches) => {
      if (err) return reject(err);
      return resolve(matches);
    })
  );

export type Args = {
  clean: boolean;
  dir: string;
  force: boolean;
  include: string[];
  exclude: string[];
};

export const handler = async ({
  clean = false,
  dir,
  exclude = [],
  force = false,
  include = []
}: Args) => {
  const CWD = process.cwd();
  const dist = path.resolve(CWD, dir);
  await ensureDir(dist);

  // Get default set of files
  const packFiles = await packlist({ path: CWD });

  // Get manually specified files
  const includedFiles: string[] = [];
  await Promise.all(
    include.map(pattern =>
      glob(pattern, { cwd: CWD }).then(matches =>
        includedFiles.push(...matches)
      )
    )
  );

  // Filter out dist dir itself
  const recursiveFilter = minimatch.filter(`!{${dir},${dir}/**}`);
  // Create new unique array from all matches and filter out dist dir
  const allFiles = Array.from(new Set(packFiles.concat(includedFiles))).filter(
    recursiveFilter
  );

  // Get final list of files after running all exclude filters
  const files = exclude.reduce(
    (files, excudeGlob) =>
      files.filter(
        (value, index, array) =>
          !minimatch.filter(excudeGlob)(value, index, array)
      ),
    allFiles
  );

  await Promise.all(
    clean
      ? // Delete all matched files when `clean` is used
        files.map((file: string) =>
          fs.promises.unlink(path.resolve(dist, file))
        )
      : // Copy all files. Throw when file exists unless `force` is used.
        files.map((file: string) =>
          fs.promises.copyFile(
            path.resolve(CWD, file),
            path.resolve(dist, file),
            force ? undefined : fs.constants.COPYFILE_EXCL
          )
        )
  );

  // Return list of files when done
  return files;
};
