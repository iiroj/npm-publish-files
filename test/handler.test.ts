import * as fs from "fs-extra";
import * as path from "path";

import handler from "../src/handler";

const TMP_DIR = path.relative(process.cwd(), "test/tmp");

test("handler", async () => {
  // Remove temp files
  await fs.remove(TMP_DIR);

  // Create temp files
  await handler({
    clean: false,
    dir: TMP_DIR,
    force: false
  });

  expect(await fs.pathExists(TMP_DIR + "/LICENSE")).toBeTruthy();
  expect(await fs.pathExists(TMP_DIR + "/package.json")).toBeTruthy();
  expect(await fs.pathExists(TMP_DIR + "/package-lock.json")).toBeFalsy();

  // Should faild to overwrite temp files without -f
  await expect(
    handler({
      clean: false,
      dir: TMP_DIR,
      force: false
    })
  ).rejects.toMatchSnapshot();

  // Succesfully overwrites temp files with -f
  await expect(
    handler({
      clean: false,
      dir: TMP_DIR,
      force: true
    })
  ).resolves.toBeUndefined();

  // Removes temp files with -c
  await handler({
    clean: true,
    dir: TMP_DIR,
    force: false
  });

  expect(await fs.pathExists(TMP_DIR + "/LICENSE")).toBeFalsy();
  expect(await fs.pathExists(TMP_DIR + "/package.json")).toBeFalsy();
});
