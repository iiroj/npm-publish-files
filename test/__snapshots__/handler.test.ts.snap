// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler 1`] = `[Error: 'test/tmp/package.json' already exists]`;
